File: python/0017_LetterCombinationsOfAPhoneNumber.py
class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """

        if '' == digits: return []
        maps = {
            '2': 'abc',
            '3': 'def',
            '4': 'ghi',
            '5': 'jkl',
            '6': 'mno',
            '7': 'pqrs',
            '8': 'tuv',
            '9': 'wxyz'
        }
        ret=['']
        for c in digits:
            tmp=[]
            for y in ret:
                for x in maps[c]:
                    tmp.append(y+x)
            ret=tmp
        
        return ret

    # def letterCombinations(self, digits):
    #     if '' == digits: return []
    #     kvmaps = {
    #         '2': 'abc',
    #         '3': 'def',
    #         '4': 'ghi',
    #         '5': 'jkl',
    #         '6': 'mno',
    #         '7': 'pqrs',
    #         '8': 'tuv',
    #         '9': 'wxyz'
    #     }
    #     return reduce(lambda acc, digit: [x + y for x in acc for y in kvmaps[digit]], digits, [''])

    

File: cpp/0017_LetterCombinationsOfAPhoneNumber.cpp
/*
通过dfs递归实现
*/
class Solution {
public:
    unordered_map<char,string> dict{{'2',"abc"},{'3',"def"},{'4',"ghi"},{'5',"jkl"},{'6',"mno"},{'7',"pqrs"},{'8',"tuv"},{'9',"wxyz"}};
    vector<string> letterCombinations(string digits) {
        vector<string> res;
        if(digits.size()==0)return res;
        dfs(res,digits,0,"");
        return res;

        
    }
    void dfs(vector<string> & res,string & digits,int pos,string s)
    {
        if(pos==digits.size()) res.push_back(s);
        else
        {
            for(int i=0;i<dict[digits[pos]].size();i++)
            {
                dfs(res,digits,pos+1,s+dict[digits[pos]][i]);
            }
        }
    }
};


